shared/src/test/scala/mlscript/ReplHost.scala

```diff
@@ -0,0 +1,193 @@
+    * @param code the main code
+    * @param res the result identifier name
+    * @return
+    */
+  def query(prelude: Str, code: Str, res: Str): ReplHost.Reply = {
+    // For empty queries, returns empty.
+    if (prelude.isEmpty && code.isEmpty)
+      ReplHost.Empty
+    else {
+      // Warp the code with `try`-`catch` block.
+      val wrapped = s"$prelude try { $code } catch (e) { console.log('\\u200B' + e + '\\u200B'); }"
+      // Send the code
+      send(wrapped)
+      // If succeed, retrieve the result.
+      parseQueryResult().map { intermediate =>
+        // Since the result might not be the result of the expression, we need
+        // to retrieve the value again.
+        send(res match {
+          case "res" => res
+          case _     => s"globalThis[\"${res}\"]"
+        })
+        parseQueryResult().map { result =>
+          // Add the intermediate result to the reply.
+          ReplHost.Result(result, Some(intermediate))
+        }
+      }
+    }
+  }
+
+  /**
+    * Execute class and function declarations.
+    *
+    * @param code the code to execute
+    * @return
+    */
+  def execute(code: Str): ReplHost.Reply = {
+    send(code)
+    collectUntilPrompt()
+  }
+
+  /**
+    * Kills the Node.js process.
+    */
+  def terminate(): Unit = proc.destroy()
+}
+
+object ReplHost {
+
+  /**
+    * The syntax error beginning text from Node.js.
+    */
+  private val syntaxErrorHead = "Uncaught SyntaxError: "
+
+  /**
+    * The base class of all kinds of REPL replies.
+    */
+  sealed abstract class Reply {
+
+    /**
+      * Maps the successful part. Like `Option[T].map`.
+      *
+      * @param f the function over
+      * @return
+      */
+    def map(f: Str => Reply): Reply
+  }
+
+  /**
+    * Represents a successful reply from Node.js.
+    *
+    * @param content the reply content, i.e. the final result
+    * @param intermediate the intermediate evaluation results
+    */
+  final case class Result(content: Str, intermediate: Opt[Str]) extends Reply {
+    override def map(f: Str => Reply): Reply = f(content)
+    override def toString(): Str = s"[success] $content"
+  }
+
+  /**
+    * If the query is `Empty`, we will receive this.
+    */
+  final object Empty extends Reply {
+    override def map(f: Str => Reply): Reply = this
+    override def toString(): Str = "[empty]"
+  }
+
+  /**
+    * If the query is `Unexecuted`, we will receive this.
+    * @param message the error message
+    */
+  final case class Unexecuted(message: Str) extends Reply {
+    override def map(f: Str => Reply): Reply = this
+    override def toString(): Str = s"[unexecuted] $message"
+  }
+
+  /**
+    * If the `ReplHost` captured errors, it will response with `Error`.
+    * @param syntax if `true`, this is a syntax error; otherwise, this is a
+    *               runtime error
+    * @param message the error message
+    */
+  final case class Error(syntax: Bool, message: Str) extends Reply {
```

Comment:
Why the special-casing for syntax errors?

If you read my message again, I suggested testing for `buffer.endsWith("\nUncaught "`. It would be better to do that and then place the end of that line (after the `Uncaught `) inside an optional error string (instead of a boolean).

This way we'll always know if something goes wrong, syntax-wise or not.

Reply:
Because we wrap code in `try`-`catch` statements, none-syntax errors will be caught by them. Code with syntax errors cannot be parsed, so `try`-`catch` statements cannot caught.

https://github.com/hkust-taco/mlscript/blob/6d5d382f7b314fecf7692a4aad2e63f416308597/shared/src/test/scala/mlscript/ReplHost.scala#L94

Reply:
Ok but are you 100% this try-catch wrapper is completely impervious to any sort of errors other than syntax errors? Does it really catch everything?

Also, is there a reason why we don't just rely on parsing `"Uncaught "` in the output to catch expressions instead of using the try/catch + `console.log('\\u200B' + e + '\\u200B')` hack?

Reply:
It's fine if there is no reason and this is just historical. Changing this is outside the scope of this PR.

Reply:
I was thinking that, compared with `"Uncaught "`, invisible characters like [ZWSP](https://en.wikipedia.org/wiki/Zero-width_space) is rarer in the output. But as we detect syntax errors by searching for `"Uncaught "`, we can remove this hack in another PR.

Reply:
Do you agree to merge this PR first and then remove this hack in the next PR?

Reply:
Yes that sounds good to me. In fact, removing the hack is low priority. Let's not do it right now – you have more urgent things to do now.